src/dune_pkg/opam_solver.mli

```diff
@@ -22,3 +22,27 @@ val solve_lock_dir
   -> Opam_repo.t list
   -> local_packages:OpamFile.OPAM.t Package_name.Map.t
   -> (Solver_result.t, [ `Diagnostic_message of _ Pp.t ]) result
+
+(** [better_candidate ~solver_env ~repos ~version_preference ~local_packages pkg] returns
+    a [Better_candidate_result.t] indicating the best version of [pkg] that can be found
+    in [repos].
+
+    It returns the following [Better_candidate_result.t]s:
+
+    - [Better_candidate {version; is_immediate_dep_of_local_package}] indicates that
+      [version] is the best available version of [pkg] and that
+      [is_immediate_dep_of_local_package] indicates whether [pkg] is a immediate
+      dependency of local package in [local_packages].
+
+    - [Package_is_best_candidate] indicates that the given pkg is already the best
+      available candidate.
+
+    - [Package_not_found] indicates that the given pkg cannot be found in any of the given
+      repositories. *)
+val better_candidate
+  :  solver_env:Solver_env.t
+  -> repos:Opam_repo.t list
+  -> version_preference:Version_preference.t
```

Comment:
Why is version preference being passed? The `Better_candidate.t` generated should be independent from this option used to direct the solver. In fact, I'm not sure why the solver environment is needed either.

Reply:
This is because we are using the `candidates` function from the `CONTEXT` we pass to 0install. Unfortunately this requires us to initialize some dummy parameters.

Version preference is still something that is relevant for us however as it suggests what a better candidate will be. I suppose this will always be newer packages.

 Solver env could probably be factored out but I didn't want to touch the 0install context code too much. Happy to do this if you prefer.

Reply:
I've removed all the usage of 0install now.